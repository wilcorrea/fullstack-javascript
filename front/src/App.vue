<template>
  <div id="app">
    <!-- where the magic happens -->
    <!-- @see front/src/router/index.ts:22 -->
    <router-view />
  </div>
</template>

<script lang="js">
import $store from 'src/store'
import Local from 'src/app/DataSource/Local'

/**
 */
export default {
  /**
   */
  name: 'App',
  /**
   */
  data: () => ({
    // state to map the version update confirmation
    confirmed: false
  }),
  /**
   */
  computed: {
    /**
     * Get the version from store
     * @see createStore
     * @returns {string}
     */
    version () {
      return $store.state.version
    }
  },
  /**
   */
  methods: {
    /**
     * This is a simple method to ask user to update the version
     * @param {string} message
     */
    notify (message) {
      if (this.confirmed) {
        return
      }
      this.confirmed = window.confirm(message)
      if (this.confirmed) {
        window.location.reload(true)
      }
    },
    /**
     * Receive the version fetched from server
     * @param {AxiosResponse} result
     */
    receiveVersion (result) {
      if (this.version && this.version !== result.data) {
        const message = `We have updates!` +
          ` The current version is ${this.version} and the new one is ${result.data}.` +
          ` Do you wanna reload the app now?`
        this.notify(message)
      }
      $store.commit('updateVersion', result.data)
    },
    /**
     * Fetch the current app version from server
     */
    fetchVersion () {
      Local
        .instance()
        .get('version')
        .then(this.receiveVersion)
    }
  },
  /**
   */
  created () {
    // wait 1 second to take a breath
    window.setTimeout(this.fetchVersion, 1000)
  }
}
</script>
