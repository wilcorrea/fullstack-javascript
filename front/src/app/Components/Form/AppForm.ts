import { CreateElement, VNode } from 'vue'
import { FormField, FormFields, Record } from 'src/app/Agnostic/interfaces'

/**
 * @component {AppForm}
 */
export default {
  /**
   */
  name: 'AppForm',
  /**
   */
  props: {
    fields: {
      type: Object,
      required: true
    },
    value: {
      type: Object,
      required: true
    }
  },
  /**
   */
  methods: {
    /**
     * @param {string} key
     * @returns {any}
     */
    getValue (key: string): any {
      // @ts-ignore
      return this.value[key]
    },
    /**
     * @param {CreateElement} h
     * @returns {Array<VNode>}
     */
    renderFields (h: CreateElement): Array<VNode> {
      // @ts-ignore
      const fields: FormFields = this.fields
      return Object.keys(fields).map((key) => {
        const field = fields[key]

        const classNames = [
          'app-field',
          `app-field--width-${field.width}`,
          `app-field--height-${field.height}`
        ]
        const data = { class: classNames }
        return h('div', data, [this.renderField(h, field)])
      })
    },
    /**
     * @param {CreateElement} h
     * @param {FormField} field
     * @returns {VNode}
     */
    renderField (h: CreateElement, field: FormField): VNode {
      const key = field.key
      const value = this.getValue(key)
      const data = {
        domProps: { value },
        props: { value },
        attrs: { ...field.attrs },
        on: this.renderFieldOn(key, field.listeners)
      }
      return h(field.is, data)
    },
    /**
     * @param {string} key
     * @param {Record} listeners
     * @returns {Record}
     */
    renderFieldOn (key: string, listeners: Record) {
      let customInput: Function
      if (listeners.input) {
        customInput = listeners.input
      }
      const input = ($event: any) => {
        // @ts-ignore
        if (customInput) {
          customInput($event)
        }
        const value = $event.target ? $event.target.value : $event
        // @ts-ignore
        this.$emit('input', { ...this.value, [key]: value })
      }
      return { ...listeners, input }
    }
  },
  /**
   * @param {CreateElement} h
   * @returns {VNode}
   */
  render (h: CreateElement): VNode {
    const form = { class: 'app-form' }
    const wrapper = { class: 'app-form__wrapper' }
    // @ts-ignore
    return h('div', form, [h('div', wrapper, [this.renderFields(h)])])
  }
}
