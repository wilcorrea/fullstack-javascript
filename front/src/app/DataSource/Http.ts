import { AxiosInstance, AxiosRequestConfig } from 'axios'

/**
 * @type {Http}
 */
export default class Http {
  /**
   * @type {string}
   */
  path = '/'

  /**
   * @type {{}}
   */
  static mocks = {}

  /**
   * @type {AxiosInstance}
   */
  private client: AxiosInstance

  /**
   * @type {Http}
   */
  private static single: Http

  /**
   * @param {AxiosInstance} client
   * @returns {this}
   */
  static instance (client?: AxiosInstance) {
    if (!this.single) {
      // @ts-ignore
      this.single = new this(client)
    }
    return this.single
  }

  /**
   * @param {AxiosInstance} client
   * @returns {this}
   */
  static build (client?: AxiosInstance) {
    // @ts-ignore
    return new this(client)
  }

  /**
   * @param {AxiosInstance} client
   */
  constructor (client: AxiosInstance) {
    this.client = client
  }

  /**
   * @param {string} url
   * @param data
   * @param {AxiosRequestConfig} config
   * @returns {Promise<any>}
   */
  post (url: string = '', data?: any, config?: AxiosRequestConfig): Promise<any> {
    const mock = Http.mockRecover(url)
    if (mock) {
      return mock
    }
    return this.client.post(url, data, config)
  }

  /**
   * @param {string} url
   * @param {AxiosRequestConfig} config
   * @returns {Promise<any>}
   */
  get (url: string = '', config?: AxiosRequestConfig): Promise<any> {
    const mock = Http.mockRecover(url)
    if (mock) {
      return mock
    }
    return this.client.get(url, config)
  }

  /**
   * @param {string} url
   * @param {AxiosRequestConfig} config
   * @returns {Promise<any>}
   */
  head (url: string, config?: AxiosRequestConfig): Promise<any> {
    const mock = Http.mockRecover(url)
    if (mock) {
      return mock
    }
    return this.client.head(url, config)
  }

  /**
   * @param {string} url
   * @param data
   * @param {AxiosRequestConfig} config
   * @returns {Promise<any>}
   */
  put (url: string, data?: any, config?: AxiosRequestConfig): Promise<any> {
    const mock = Http.mockRecover(url)
    if (mock) {
      return mock
    }
    return this.client.put(url, data, config)
  }

  /**
   * @param {string} url
   * @param data
   * @param {AxiosRequestConfig} config
   * @returns {Promise<any>}
   */
  patch (url: string, data?: any, config?: AxiosRequestConfig): Promise<any> {
    const mock = Http.mockRecover(url)
    if (mock) {
      return mock
    }
    return this.client.patch(url, data, config)
  }

  /**
   * @param {string} url
   * @param {AxiosRequestConfig} config
   * @returns {Promise<any>}
   */
  delete (url: string, config?: AxiosRequestConfig): Promise<any> {
    const mock = Http.mockRecover(url)
    if (mock) {
      return mock
    }
    return this.client.delete(url, config)
  }

  /**
   * @param {string} path
   * @param value
   */
  static mockRegister (path: string, value: any = null): void {
    // @ts-ignore
    Http.mocks[path] = value
  }

  /**
   * @param {string} path
   * @returns {any}
   */
  static mockRecover (path: string): any {
    // @ts-ignore
    return Http.mocks[path]
  }
}
