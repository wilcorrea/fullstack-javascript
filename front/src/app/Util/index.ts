import { get, clone } from './general'

export default {
  get, clone
}
