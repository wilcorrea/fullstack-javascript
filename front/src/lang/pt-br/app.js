// the app built-in texts
export default {
  components: {
    pagination: {
      more: 'Mais resultados'
    }
  }
}
