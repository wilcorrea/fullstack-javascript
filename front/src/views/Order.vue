<template>
  <div class="app-layout__page Order">
    <h2>{{ $lang('views.order.title') }}</h2>
    <!-- @see front/src/app/Components/Form/AppForm.ts -->
    <app-form
      v-bind="form"
      v-model="record"
    />
  </div>
</template>

<script lang="js">
import OrderSchema from 'src/domains/Order/OrderSchema'
import OrderService from 'src/domains/Order/OrderService'
import HooksMixin from 'src/domains/Common/HooksMixin'

/**
 * @component {Order}
 */
export default {
  /**
   */
  name: 'Order',
  /**
   * Mixin usage to handle with schema hooks
   */
  mixins: [HooksMixin],
  /**
   */
  props: {
    // the id of item to be fetched
    id: {
      type: [String, Number],
      default: undefined
    }
  },
  /**
   */
  data: () => ({
    // the form states
    form: {
      fields: {}
    },
    // the record to show
    record: {}
  }),
  /**
   */
  methods: {
    /**
     * Fetch a record from service
     * @params {string} id
     */
    fetchRecord (id) {
      OrderService
        .instance()
        .read(id)
        .then(this.assignRecord)
    },
    /**
     * Handle with the record fetched
     * @param {Object} result
     */
    assignRecord (result) {
      if (result && typeof result.order === 'object') {
        this.$options.payload = result.order
        Object.keys(this.record).forEach(this.setRecordProperty)
        this.triggerHook('fetch:record')
      }
    },
    /**
     * Set the record property with the payload
     * All the data is store in payload, payload is not in data because it is not reactive
     * @param {string} key
     */
    setRecordProperty (key) {
      const value = this.$util.get(this.$options.payload, key)
      this.$set(this.record, key, value)
      // this.record[key] = this.$util.get(this.$options.payload, key)
    },
    /**
     * Start record with the default values
     * @param {Object} record
     */
    setRecord (record) {
      this.record = record
    },
    /**
     * Perform the created life cycle hook
     */
    onCreated () {
      const { fields, record, hooks } = OrderSchema.instance().form()
      this.form.fields = fields
      this.setRecord(record)
      this.installHooks(hooks)
    },
    /**
     * Perform the mounted life cycle hook
     */
    onMounted () {
      // try to get id from props
      // used in tests
      let id = this.id
      if (this.$route) {
        // the route params have higher priority
        id = this.$route.params.id
      }
      if (!id) {
        return
      }
      // fetch the record
      this.fetchRecord(id)
    }
  },
  /**
   */
  created () {
    this.onCreated()
  },
  /**
   */
  mounted () {
    this.onMounted()
  }
}
</script>
