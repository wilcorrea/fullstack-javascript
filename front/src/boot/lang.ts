import $lang from 'src/lang'

/**
 * @param {Vue} Vue
 * @returns {}
 */
export default (Vue: any) => {
  /**
   */
  Object.defineProperty(Vue.prototype, '$lang', {
    get () {
      return $lang
    }
  })
}
