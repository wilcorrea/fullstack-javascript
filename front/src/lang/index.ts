import messages from './messages'
import { get } from 'src/app/Util/general'
import { i18n } from 'src/settings'

/**
 * Start with a default locale
 * @type {string}
 */
let locale = i18n()

/**
 * Set the current locale
 * @param {string} newLocale
 */
export const setLocale = (newLocale: string) => {
  locale = newLocale
}

/**
 * Get the current locale
 * @returns {string}
 */
export const getLocale = (): string => {
  return locale
}

/**
 * Lang is a featured i18n engine
 * We can change the locale and add more messages in realtime
 * I do not like some things in vue-i18n and other libs that I've tested, so that's it
 * @param {string} path
 * @param {string} fallback
 * @returns {*}
 */
export default (path: string, fallback: string = ''): any => {
  return get(messages(locale), path, fallback)
}
