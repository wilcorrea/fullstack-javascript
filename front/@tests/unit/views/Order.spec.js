import { shallowMount } from '@vue/test-utils'
import Order from 'src/views/Order.vue'
import $lang from 'src/lang'
import $util from 'src/app/Util'
import Http from 'src/app/DataSource/Http'
import AppForm from 'src/app/Components/Form/AppForm'

describe('Order', () => {
  const mock = { order: { id: 1 } }
  Http.mockRegister(`/orders/get/${mock.order.id}`, mock)

  const wrapper = shallowMount(
    Order,
    {
      propsData: mock.order,
      mocks: { $lang, $util },
      components: { AppForm }
    }
  )

  /**
   * @test
   */
  it('check component markup', () => {
    expect(wrapper.html()).toContain(`<h2>${$lang('views.order.title')}</h2>`)
  })

  /**
   * @test
   */
  it('method assignRecord', () => {
    wrapper.vm.assignRecord(mock)
    expect(mock.order.id).toEqual(wrapper.vm.record.id)
  })

  /**
   * @test
   */
  it('method fetchRecord', async () => {
    await wrapper.vm.fetchRecord(mock.order.id)
    expect(mock.order).toEqual(wrapper.vm.record)
  })

  /**
   * @test
   */
  test('should render content correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
