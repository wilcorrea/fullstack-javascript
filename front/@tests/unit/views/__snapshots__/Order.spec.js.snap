// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order should render content correctly 1`] = `
<div class="app-layout__page Order">
  <h2>Detalhes do Pedido</h2>
  <app-form-stub fields="[object Object]" value="[object Object]"></app-form-stub>
</div>
`;
